<h1> Homework 7 </h1>
<h2> James Roberts - Math 104 </h2>

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv("personality20_labeled_oriented.csv")
df

A_q13  A_q2  A_q6  A_q9  C_q15  C_q20  C_q5  C_q8  E_q10  E_q14  E_q18  \
0         1    -2     2    -3      4     -2     5    -2      5     -1     -1   
1         2    -2     3    -3      4     -3     4    -4      4     -1     -1   
2         2    -2     4    -3      2     -4     4    -5      4     -1     -1   
3         2    -5     2    -4      3     -2     2    -4      3     -4     -2   
4         3    -2     5    -2      4     -2     4    -4      3     -3     -4   
...     ...   ...   ...   ...    ...    ...   ...   ...    ...    ...    ...   
9995      3    -1     3    -1      4     -1     5    -2      3     -3     -1   
9996      2    -4     2    -5      3     -2     3    -5      1     -5     -4   
9997      5    -5     1    -3      2     -3     4    -5      4     -5     -1   
9998      3    -3     5    -1      2     -1     5    -2      4     -4     -1   
9999      2    -4     3    -4      2     -2     4    -5      3     -3     -4   

      E_q3  N_q11  N_q12  N_q17  N_q19  O_q1  O_q16  O_q4  O_q7  
0        5     -2      3      3     -2     5      5    -1    -3  
1        4     -4      2      4     -2     5      4    -3    -1  
2        5     -4      3      4     -2     4      4    -3    -1  
3        4     -3      4      4     -2     4      2    -2    -2  
4        3     -4      5      3     -4     5      5    -1    -1  
...    ...    ...    ...    ...    ...   ...    ...   ...   ...  
9995     2     -4      2      2     -4     4      3    -4    -2  
9996     2     -1      5      5     -2     5      4    -2    -2  
9997     4     -3      4      4     -5     4      5    -4    -3  
9998     2     -5      1      5     -2     4      4    -2    -3  
9999     4     -2      4      3     -4     4      4    -3    -3  

[10000 rows x 20 columns]

In [3]:
# Get the mean of each column
df_means = df.mean()
df_means

A_q13    3.5979
A_q2    -2.8718
A_q6     3.6695
A_q9    -2.0533
C_q15    3.4809
C_q20   -2.0794
C_q5     4.1437
C_q8    -3.0098
E_q10    3.6473
E_q14   -2.9559
E_q18   -2.2714
E_q3     3.3689
N_q11   -3.6997
N_q12    2.7000
N_q17    3.0840
N_q19   -3.1488
O_q1     4.3350
O_q16    3.7919
O_q4    -2.3303
O_q7    -1.9772
dtype: float64

In [4]:
# Demean each column
df = df.sub(df_means, axis = 'columns')
df

A_q13    A_q2    A_q6    A_q9   C_q15   C_q20    C_q5    C_q8   E_q10  \
0    -2.5979  0.8718 -1.6695 -0.9467  0.5191  0.0794  0.8563  1.0098  1.3527   
1    -1.5979  0.8718 -0.6695 -0.9467  0.5191 -0.9206 -0.1437 -0.9902  0.3527   
2    -1.5979  0.8718  0.3305 -0.9467 -1.4809 -1.9206 -0.1437 -1.9902  0.3527   
3    -1.5979 -2.1282 -1.6695 -1.9467 -0.4809  0.0794 -2.1437 -0.9902 -0.6473   
4    -0.5979  0.8718  1.3305  0.0533  0.5191  0.0794 -0.1437 -0.9902 -0.6473   
...      ...     ...     ...     ...     ...     ...     ...     ...     ...   
9995 -0.5979  1.8718 -0.6695  1.0533  0.5191  1.0794  0.8563  1.0098 -0.6473   
9996 -1.5979 -1.1282 -1.6695 -2.9467 -0.4809  0.0794 -1.1437 -1.9902 -2.6473   
9997  1.4021 -2.1282 -2.6695 -0.9467 -1.4809 -0.9206 -0.1437 -1.9902  0.3527   
9998 -0.5979 -0.1282  1.3305  1.0533 -1.4809  1.0794  0.8563  1.0098  0.3527   
9999 -1.5979 -1.1282 -0.6695 -1.9467 -1.4809  0.0794 -0.1437 -1.9902 -0.6473   

       E_q14   E_q18    E_q3   N_q11  N_q12  N_q17   N_q19   O_q1   O_q16  \
0     1.9559  1.2714  1.6311  1.6997    0.3 -0.084  1.1488  0.665  1.2081   
1     1.9559  1.2714  0.6311 -0.3003   -0.7  0.916  1.1488  0.665  0.2081   
2     1.9559  1.2714  1.6311 -0.3003    0.3  0.916  1.1488 -0.335  0.2081   
3    -1.0441  0.2714  0.6311  0.6997    1.3  0.916  1.1488 -0.335 -1.7919   
4    -0.0441 -1.7286 -0.3689 -0.3003    2.3 -0.084 -0.8512  0.665  1.2081   
...      ...     ...     ...     ...    ...    ...     ...    ...     ...   
9995 -0.0441  1.2714 -1.3689 -0.3003   -0.7 -1.084 -0.8512 -0.335 -0.7919   
9996 -2.0441 -1.7286 -1.3689  2.6997    2.3  1.916  1.1488  0.665  0.2081   
9997 -2.0441  1.2714  0.6311  0.6997    1.3  0.916 -1.8512 -0.335  1.2081   
9998 -1.0441  1.2714 -1.3689 -1.3003   -1.7  1.916  1.1488 -0.335  0.2081   
9999 -0.0441 -1.7286  0.6311  1.6997    1.3 -0.084 -0.8512 -0.335  0.2081   

        O_q4    O_q7  
0     1.3303 -1.0228  
1    -0.6697  0.9772  
2    -0.6697  0.9772  
3     0.3303 -0.0228  
4     1.3303  0.9772  
...      ...     ...  
9995 -1.6697 -0.0228  
9996  0.3303 -0.0228  
9997 -1.6697 -1.0228  
9998  0.3303 -1.0228  
9999 -0.6697 -1.0228  

[10000 rows x 20 columns]

In [5]:
# All columns now have a mean of approximately 0
df.mean()

A_q13   -9.379164e-17
A_q2    -2.273737e-17
A_q6    -1.968203e-16
A_q9     1.250555e-16
C_q15   -1.051603e-16
C_q20    8.668621e-17
C_q5     6.181722e-17
C_q8    -1.250555e-16
E_q10    7.105427e-17
E_q14    1.371347e-16
E_q18   -1.364242e-16
E_q3    -5.684342e-18
N_q11   -1.136868e-17
N_q12   -1.620037e-16
N_q17    0.000000e+00
N_q19   -2.984279e-17
O_q1     3.552714e-17
O_q16   -4.831691e-17
O_q4    -2.529532e-16
O_q7    -3.126388e-17
dtype: float64

In [6]:
# Convert Dataframe to NumPy Array
A = np.array(df)
A

array([[-2.5979,  0.8718, -1.6695, ...,  1.2081,  1.3303, -1.0228],
       [-1.5979,  0.8718, -0.6695, ...,  0.2081, -0.6697,  0.9772],
       [-1.5979,  0.8718,  0.3305, ...,  0.2081, -0.6697,  0.9772],
       ...,
       [ 1.4021, -2.1282, -2.6695, ...,  1.2081, -1.6697, -1.0228],
       [-0.5979, -0.1282,  1.3305, ...,  0.2081,  0.3303, -1.0228],
       [-1.5979, -1.1282, -0.6695, ...,  0.2081, -0.6697, -1.0228]])

In [7]:
U,S,V = np.linalg.svd(A)

In [8]:
# Singular values of A
S

array([226.87400051, 168.82039569, 157.13143628, 137.89677347,
       131.79179584, 112.28273434, 101.71606183,  98.49674646,
        96.53580717,  93.61519741,  90.41274137,  89.11558621,
        86.34058548,  83.9091372 ,  83.18041094,  80.31469778,
        74.42541195,  72.84108652,  69.74028457,  64.17977628])

The first singular value of 226.87 is significantly larger than the others (the next largest singular value is 168.82). This means that this variable has more influence and accounts for more variance than the other variables. We do not need all of the questions to explain the variability in the data. 

There are a few other singular values that may help to account for more of the variability, but are not entirely necessary. We likely do not need all 5 traits of OCEAN to estimate the personality of someone because the large first singular value is able to explain the majority of the variability in the data.
